### PR TITLE
Update base version to 0.18.x

### PR DESCRIPTION
**Pull request by cryptcoin-junkey**
Just rebased master-0.17 onto litecoin/0.18. It have not been ~~built and~~ tested yet.

### CODE REVIEW

**Comment by cryptcoin-junkey** [comment]
[FYI] What I did.

```
git checkout -b develop-0.18 litecoin/0.18
git rebase --onto HEAD 1b6c48  master-0.17
{loop}
vi {conflicted files}
git add --all
git rebase --continue
{/loop until all conflicts are resolved}
```

I suppose this way is one of the easiest way.

**Comment by cryptcoin-junkey** [comment]
[FYI] I'll push my commits with `--force` into this PR branch until it was merged.
Please have in mind if you track this.

**Comment by cryptcoin-junkey** [comment]
The minimum build (monacoind with '--without-wallet`) was passed.
QT has not been tried yet.

Of course, this doesn't mean "binaries from this tree are stable."

**Comment by cryptcoin-junkey** [comment]
I'm trying full build and found build errors. I'm going to fix them. The workaround is `configure --without-wallet` for now.

**Comment by wakiyamap** [comment]
If I find a better place to change, should I write it here?

```
mv doc/man/litecoin-wallet.1 doc/man/monacoin-wallet.1
mv doc/litecoin-conf.md doc/monacoin-conf.md
```

**Comment by cryptcoin-junkey** [comment]
@wakiyamap Nice catch. I did it and pushed.

**Comment by cryptcoin-junkey** [comment]
I force pushed commits here. I believe `monacoin-wallet` can be built now.

File: src/policy/feerate.cpp
Comment (wakiyamap) [inline src/policy/feerate.cpp:13]: If value of `nFeePaid` is 10512000000000000(MAX_MONEY), it's overflow?
nFeePaid * 1000 = 
`10,512,000,000,000,000,000`
int64_t of max is 
`9,223,372,036,854,775,807`

So......how about adding the following?
`assert(nFeePaid<= MAX_MONEY);`

Personaly, I'd like to believe that maybe this isn't the case.

**Comment by wakiyamap** [comment]
I could build it. yay!!

File: doc/man/monacoin-wallet.1
Comment (wakiyamap) [inline doc/man/monacoin-wallet.1]: Monatcoin -> Monacoin

**Comment by wakiyamap** [comment]
Litecoin -> Monacoin
src/config/bitcoin-config.h
```
/* Define to the full name of this package. */
#define PACKAGE_NAME "Monacoin Core"

/* Define to the full name and version of this package. */
#define PACKAGE_STRING "Monacoin Core 0.18.1"
```

src/Makefile
```
PACKAGE_NAME = Monacoin Core
PACKAGE_STRING = Monacoin Core 0.18.1
```
image

File: doc/man/monacoin-wallet.1
Comment (cryptcoin-junkey) [reply doc/man/monacoin-wallet.1]: Nice catch! I fixed it.

**Comment by cryptcoin-junkey** [comment]
@wakiyamap I see. They should be fixed `AC_INIT` in `configure.ac`. I did it and pushed here.

File: src/policy/feerate.cpp
Comment (yamada-guro-baru) [reply src/policy/feerate.cpp:13]: x  `assert(nFeePaid <= MAX_MONEY);`
o  `assert(nFeePaid <= INT64_MAX / 1000);`

`MAX_MONEY` is niet goed genoeg.

**Comment by wakiyamap** [comment]
There is a python module called monacoin_scrypt in the source.
Should I have monacoin_scrypt in pypi?
Ex.. https://pypi.org/project/litecoin_scrypt/

**Comment by yamada-guro-baru** [comment]
@wakiyamap 
lyra2re2_hash, I thought.......
But, If we're just testing, maybe scrypt is the way to go.

**Comment by cryptcoin-junkey** [comment]
@wakiyamap I grepped in the branch.
And I found `monacoin_scrypt` related code in `.travis.yml` and `test/functional/test_framework/messages.py'

In `.travis.yml`, we don't need to care because `$MONACOIN_SCRYPT` won't set to `1`.

IMO, `message.py` will be useless now even if the PIP module was applied. Totally tests have been broken.

I agree that tests should be refactored for Monacoin. Someday we could do it.
But code in this branch will have similar quality to the released code even if we didn't refactor.

File: src/policy/feerate.cpp
Comment (cryptcoin-junkey) [reply src/policy/feerate.cpp:13]: Is this a good practice to use `asset()` here?
I suppose all Monacoin-core nodes will be aborted by assertion failed.
`assert()` in current code is checks for the runtime (compiler) limitations. Not for argument checks.

My alternative plan is to use `int128_t` and/or `uint128_t` in functions. WDYT?

**Comment by wakiyamap** [comment]
@cryptcoin-junkey 
ok. I agree. I think it's not necessary to all test it ""now"".
This is fine for the base(0.18.x).

But if we're going to release it, please just add `maxfeerate`.

File: src/policy/feerate.cpp
Comment (wakiyamap) [reply src/policy/feerate.cpp:13]: It's good.
I'm not very familiar with C++, but it looks like it says you have to use `BOOST`.
If you're putting in a new library, so......is it better to use `arith_uint256`?

Comment (yamada-guro-baru) [reply src/policy/feerate.cpp:13]: ...fm
nFeePaid * 1000 = 10,512,000,000,000,000,000
max of uint64_t   = 18,446,744,073,709,551,615
so,
```
    uint64_t uFeePaid = std::abs(nFeePaid) * 1000;
    if (nSize > 0 && nFeePaid > 0) {
        nSatoshisPerK = CAmount(uFeePaid / nSize);
    } else if (nSize > 0 && nFeePaid < 0) {
        nSatoshisPerK = CAmount(-uFeePaid / nSize);
    } else {
        nSatoshisPerK = 0;
    }
```
How do you think?

Comment (cryptcoin-junkey) [reply src/policy/feerate.cpp:13]: How about fixing the type of `CAmount` ?

In `src/amount.h`:

```diff
  /** Amount in satoshis (Can be negative) */
- typedef int64_t CAmount;
+ typedef int128_t CAmount;
```

It looks this was discussed in [the bitcoin core team](http://gnusha.org/bitcoin-core-dev/2019-07-15.log).

> 11:16 < Raystonn> Normally I'd agree.  But upon testing a change of CAmount to  boost::multiprecision::int128_t, the floating point operations aren't even supported.  I could add them, but it could lead to dropped precision or accuracy some place where it matters simply by defining the operators.
> 11:16 < sipa> there is no need for 128 bit integers
> 11:16 < Raystonn> Right now no.

They didn't need it. And it may be worth to consider for us.

Comment (cryptcoin-junkey) [reply src/policy/feerate.cpp:13]: Ooops, `s/int128_t/__int128_t/` . AFAIK, `__int128_t` is GCC extension, supported by Clang also.

Comment (cryptcoin-junkey) [reply src/policy/feerate.cpp:13]: Another plan. I guess this plan is safer.

```diff
-         nSatoshisPerK = nFeePaid * 1000 / nSize;
+         nSatoshisPerK = (CAmount)((__int128_t)nFeePaid * 1000 / nSize);
```

Comment (wakiyamap) [reply src/policy/feerate.cpp:13]: @cryptcoin-junkey 
I disagree with setting CAmount to `int128_t`.
The impact is toooooooooo great.

I agree to an another plan.

**Comment by wakiyamap** [comment]
[English]
Why not decide on the criteria to release?
The overflow fix is complete, etc......

My suggestion.
As soon as the overflow fix is complete, run it on your own node for a month or so and release it if it's OK.

[日本語]
リリースする基準を決めませんか？
オーバーフローの修正が完了したら等……

私としてはオーバーフローの修正が完了して、1か月ほど各自でノードを運用して問題なければリリースで良いかなと思います。

**Comment by cryptcoin-junkey** [comment]
I pushed the fix against #76.

**Review by wakiyamap — APPROVED** [review]

**Pull request merged**
